internal/vfs/pnpvfs/pnpvfs.go

```diff
@@ -0,0 +1,399 @@
+			dir, ok := dirs[lowerDir]
+			if !ok {
+				dir = &compressedDir{
+					path:    dirPath,
+					entries: make(map[string]EntryKind),
+				}
+
+				// List the same directory both with and without the slash
+				dirs[lowerDir] = dir
+				dirs[lowerDir+"/"] = dir
+			}
+			dir.entries[baseName] = DirEntry
+			baseName = dirPath
+		}
+	}
+
+	archive.dirs = dirs
+	archive.files = files
+	archive.reader = reader
+}
+
+func (fs *PNPFS) UseCaseSensitiveFileNames() bool {
+	return fs.inner.UseCaseSensitiveFileNames()
+}
+
+func (fs *PNPFS) FileExists(path string) bool {
+	path = mangleYarnPnPVirtualPath(path)
+
+	if fs.inner.FileExists(path) {
+		return true
+	}
+
+	zip, pathTail := fs.checkForZip(path, FileEntry)
+	if zip == nil {
+		return false
+	}
+
+	_, ok := zip.files[strings.ToLower(pathTail)]
+	return ok
+}
+
+func (fs *PNPFS) ReadFile(path string) (contents string, ok bool) {
+	path = mangleYarnPnPVirtualPath(path)
+
+	contents, ok = fs.inner.ReadFile(path)
+	if ok {
+		return contents, ok
+	}
+
+	// If the file doesn't exist, try reading from an enclosing zip archive
+	zip, pathTail := fs.checkForZip(path, FileEntry)
+	if zip == nil {
+		return "", false
+	}
+
+	// Does the zip archive have this file?
+	file, ok := zip.files[strings.ToLower(pathTail)]
+	if !ok {
+		return "", false
+	}
+
+	// Check whether it has already been read
+	file.mutex.Lock()
+	defer file.mutex.Unlock()
+
+	if file.wasRead {
+		return file.contents, file.err == nil
+	}
+	file.wasRead = true
+
+	// If not, try to open it
+	reader, err := file.compressed.Open()
+	if err != nil {
+		file.err = err
+		return "", err == nil
+	}
+	defer reader.Close()
+
+	// Then try to read it
+	bytes, err := io.ReadAll(reader)
+	if err != nil {
+		file.err = err
+		return "", false
+	}
+
+	file.contents = string(bytes)
+
+	return file.contents, true
+}
+
+func (fs *PNPFS) WriteFile(path string, data string, writeByteOrderMark bool) error {
+	return fs.inner.WriteFile(path, data, writeByteOrderMark)
+}
+
+func (fs *PNPFS) Remove(path string) error {
+	return fs.inner.Remove(path)
+}
+
+func (fs *PNPFS) Chtimes(path string, aTime time.Time, mTime time.Time) error {
+	return fs.inner.Chtimes(path, aTime, mTime)
+}
+
+func (fs *PNPFS) DirectoryExists(path string) bool {
+	path = mangleYarnPnPVirtualPath(path)
+
+	if fs.inner.DirectoryExists(path) {
+		return true
+	}
+
+	zip, pathTail := fs.checkForZip(path, DirEntry)
+	if zip == nil {
+		return false
+	}
+
+	_, ok := zip.dirs[strings.ToLower(pathTail)]
+	return ok
+}
+
+func (fs *PNPFS) GetAccessibleEntries(path string) vfs.Entries {
+	path = mangleYarnPnPVirtualPath(path)
+
+	entries := fs.inner.GetAccessibleEntries(path)
+	if len(entries.Files) > 0 || len(entries.Directories) > 0 {
+		return entries
+	}
+
+	zip, pathTail := fs.checkForZip(path, DirEntry)
+	if zip == nil {
+		return entries
+	}
+
+	dir, ok := zip.dirs[strings.ToLower(pathTail)]
+	if !ok {
+		return entries
+	}
+
+	files := make([]string, 0)
+	dirs := make([]string, 0)
```

Comment:
`make([]string, 0)` is just an allocate-y version of `var files []string`; don't intentionally make an empty slice unless `nil` vs non-`nil` matters.